### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by KonstantinosTsivis**
I reformatted the code for news.py plugin. Some of the functionalities I included are:
- cleared previous code which had very high coupling
- gave the user the option to search based on a specific keyword
- included news <help> menu so that the user knows the differences between each command they can use
- deleted repeated code
- gave clear output to the user when it comes to configuring the own news channel. Previously, they had to manually enter their API to the memory.json file, and the proccess of how to get the API key and how to do that was not clear. Now, they can just enter it to the command line, and it gets added automatically. Also I have the option to update it.
- option to remove a news source from their possible list of news sources

### CODE REVIEW

**Comment by appi147** [comment]
pep8 errors:
```
jarviscli/plugins/news.py:23:30: E226 missing whitespace around arithmetic operator
jarviscli/plugins/news.py:204:1: W391 blank line at end of file
```

File: jarviscli/plugins/news.py [hunk 1]
```diff
@@ -1,224 +1,157 @@
 # !!! This uses the https://newsapi.org/ api. TO comply with the TOU
 # !!! we must link back to this site whenever we display results.
+import json
+import webbrowser
+from six.moves import input
+from plugin import plugin, require
+from colorama import Fore
 try:  # python3
     import urllib.request
     import urllib.parse
```
Comment (appi147) [inline jarviscli/plugins/news.py]: line 10 and 11 imports are not needed I guess?

**Comment by appi147** [comment]
Have you tested all the changes

**Comment by appi147** [comment]
```
~> Hi, what can I do for you?
news
Some error occurred, please open an issue on github!
Here is error:

Traceback (most recent call last):
  File "jarviscli/CmdInterpreter.py", line 155, in try_do
    do(self, s)
  File "jarviscli/plugin.py", line 219, in _plugin_run_with_network_error
    run_func(jarvis, s)
  File "jarviscli/plugin.py", line 201, in run
    self._backend[0](jarvis.get_api(), s)
  File "/home/appi/Public/github/Jarvis/jarviscli/plugins/news.py", line 54, in __call__
    self.get_headlines(jarvis)
  File "/home/appi/Public/github/Jarvis/jarviscli/plugins/news.py", line 131, in get_headlines
    url = "https://newsapi.org/v2/top-headlines?country=us&apiKey=" + self.get_api_key(jarvis)
TypeError: can only concatenate str (not "NoneType") to str


```

**Comment by appi147** [comment]
> ```
> ~> Hi, what can I do for you?
> news
> Some error occurred, please open an issue on github!
> Here is error:
> 
> Traceback (most recent call last):
>   File "jarviscli/CmdInterpreter.py", line 155, in try_do
>     do(self, s)
>   File "jarviscli/plugin.py", line 219, in _plugin_run_with_network_error
>     run_func(jarvis, s)
>   File "jarviscli/plugin.py", line 201, in run
>     self._backend[0](jarvis.get_api(), s)
>   File "/home/appi/Public/github/Jarvis/jarviscli/plugins/news.py", line 54, in __call__
>     self.get_headlines(jarvis)
>   File "/home/appi/Public/github/Jarvis/jarviscli/plugins/news.py", line 131, in get_headlines
>     url = "https://newsapi.org/v2/top-headlines?country=us&apiKey=" + self.get_api_key(jarvis)
> TypeError: can only concatenate str (not "NoneType") to str
> ```

May be add some custom error messages to add apikey to memory.json

**Comment by appi147** [comment]
Please write doc comments after method names

**Comment by appi147** [comment]
I liked the option to search for news for a specific term. But when more than one term are given it crashes.
e.g. 
news google -- it works but all the following causes a crash
news google stadia
news "google stadia"

Error:
```
Some error occurred, please open an issue on github!
Here is error:

Traceback (most recent call last):
  File "jarviscli/CmdInterpreter.py", line 155, in try_do
    do(self, s)
  File "jarviscli/plugin.py", line 219, in _plugin_run_with_network_error
    run_func(jarvis, s)
  File "jarviscli/plugin.py", line 201, in run
    self._backend[0](jarvis.get_api(), s)
  File "/home/appi/Public/github/Jarvis/jarviscli/plugins/news.py", line 56, in __call__
    self.get_news(jarvis, s)
  File "/home/appi/Public/github/Jarvis/jarviscli/plugins/news.py", line 152, in get_news
    return self.parse_articles(url)
  File "/home/appi/Public/github/Jarvis/jarviscli/plugins/news.py", line 159, in parse_articles
    response = urllib.request.urlopen(url)
  File "/usr/lib/python3.7/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.7/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/usr/lib/python3.7/urllib/request.py", line 641, in http_response
    'http', request, response, code, msg, hdrs)
  File "/usr/lib/python3.7/urllib/request.py", line 569, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.7/urllib/request.py", line 503, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.7/urllib/request.py", line 649, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 400: Bad Request
```

**Comment by appi147** [comment]
In help displayed (see line 51), `removesource` is displayed as a method to remove sources, but as seen in line 44 `remove` term is used, leading to removesearch as a search term in `news [word]`

File: jarviscli/plugins/news.py [hunk 2]
```diff
@@ -1,224 +1,157 @@
...
-                get_configured_opt(jarvis)
-            # if not set get users preference
-            else:
-                self.get_default_opt(jarvis)
-        elif self.is_news_site_configured:
-            self.get_configured_opt(jarvis)
+    """
+        configures the news channel of the user
+    """
+    def configure(self, jarvis):
```
Comment (appi147) [inline jarviscli/plugins/news.py]: In configure function, IMO there is no need of choice if a user has entered `news configure`. Maybe give an index to not make any changes

File: jarviscli/plugins/news.py [hunk 3]
```diff
@@ -1,224 +1,157 @@
...
+            if key.lower() == "q" or key.lower() == "":
+                jarvis.say("Could not update the NEWS API key! ", Fore.RED)
+            else:
+                self.update_api_key(jarvis, key)
+                jarvis.say("NEWS API key successfully updated! ", Fore.GREEN)
+        elif s == "configure":
+            self.configure(jarvis)
+        elif s == "remove":
+            source_to_remove = input("Please enter the news source you want to remove: ")
+            self.remove_source(jarvis, source_to_remove)
```
Comment (appi147) [inline jarviscli/plugins/news.py]: Maybe display list of current sources, as implemented in configure

File: jarviscli/plugins/news.py [hunk 4]
```diff
@@ -1,224 +1,157 @@
...
+        choice = input('Would you like to configure your own news channel(y/n) ')
+        if choice == 'y' or choice == 'yes':
+            for index in self.source_dict.keys():
+                print(str(index) + ": " + self.source_dict.get(index))
+            index_list = input("Type the indexes of the sources you would like to add to your channel separated by "
+                               "space: ")
+            index_list = index_list.split(" ")
+            for index in index_list:
+                self.add_source(jarvis, self.source_dict.get(index, index))
+            print("Visit https://newsapi.org/sources to add any sources not in the previous list")
```
Comment (appi147) [inline jarviscli/plugins/news.py]: I dont understand, isn't all the sources hardcoded

File: jarviscli/plugins/news.py [hunk 5]
```diff
@@ -1,224 +1,157 @@
...
 @plugin('news')
-class News():
-    """
-    Time to get an update about the local news.
-    Type \"news\" to choose your source or \"news quick\" for some headlines.
-    """
-    def __init__(self, source="google-news", api_key="7488ba8ff8dc43459d36f06e7141c9e5"):
+class News:
+
+    def __init__(self, api_key="7488ba8ff8dc43459d36f06e7141c9e5"):
```
Comment (appi147) [inline jarviscli/plugins/news.py]:
```suggestion
    def __init__(self, api_key):
```

File: jarviscli/plugins/news.py [hunk 6]
```diff
@@ -1,224 +1,157 @@
...
         self.apiKey = api_key
```
Comment (appi147) [inline jarviscli/plugins/news.py]:
```suggestion
```

File: jarviscli/plugins/news.py [hunk 7]
```diff
@@ -1,224 +1,157 @@
...
+            url += "&sources="
+            for source in sources:
+                url += source + ","
+        url += "&apiKey=" + self.get_api_key(jarvis)
+        return self.parse_articles(url)
 
-    def request_news(self, url=None):
+    def parse_articles(self, url):
         # check to see if a url was passed
         if url is None:
```
Comment (appi147) [inline jarviscli/plugins/news.py]:
```suggestion
```

File: jarviscli/plugins/news.py [hunk 8]
```diff
@@ -1,224 +1,157 @@
...
             url = self.url
```
Comment (appi147) [inline jarviscli/plugins/news.py]:
```suggestion
```

File: jarviscli/plugins/news.py [hunk 9]
```diff
@@ -1,224 +1,157 @@
...
-    Time to get an update about the local news.
-    Type \"news\" to choose your source or \"news quick\" for some headlines.
-    """
-    def __init__(self, source="google-news", api_key="7488ba8ff8dc43459d36f06e7141c9e5"):
+class News:
+
+    def __init__(self, api_key="7488ba8ff8dc43459d36f06e7141c9e5"):
         self.apiKey = api_key
-        self.source = source
         self.url = "https://newsapi.org/v1/articles?source=google-news&sortBy=top" \
```
Comment (appi147) [inline jarviscli/plugins/news.py]:
```suggestion
```

File: jarviscli/plugins/news.py [hunk 10]
```diff
@@ -1,224 +1,157 @@
...
-                   "&apiKey=7488ba8ff8dc43459d36f06e7141c9e5"
+                   "&apiKey="+self.apiKey
```
Comment (appi147) [inline jarviscli/plugins/news.py]:
```suggestion
```

**Comment by appi147** [comment]
Okay this looks overwhelming. I didn't mean to scare you. I will summarise all the above review:

- These are pep8 errors

> jarviscli/plugins/news.py:23:30: E226 missing whitespace around arithmetic operator
> jarviscli/plugins/news.py:204:1: W391 blank line at end of file

- In line 20 api_key is passed and self.apiKey is defined an self.url is created, which are not used anywhere. I suggest to remove these as apikey is always fetched from memory.json

- In help removesource is displayed whereas actual command is remove

- While removing sources, program says to manually input sources, which a user doesn't know exact name of sources. Maybe display a list of current sources and remove by index as implemented in configure method

- I appreciate the option to search for a specific term, but maybe there was an option to search for two terms [https://github.com/sukeesh/Jarvis/pull/439#issuecomment-475524310](https://github.com/sukeesh/Jarvis/pull/439#issuecomment-475524310) e.g. `news google stadia`

- some of the imports are unused maybe such as urslib.parse, urllib.error.

- A choice has been asked about whether to configure or not after a user has entered `news configure`. It can be simplified.

- In line 156 an if statement is added if no url  is passed. It is not needed as it is internal function, url is always passed, don't want to use self.apiKey or self.url. 

- In line 119, a text is displayed about updating sources online. I don't  understand this as sources are hard-coded in line 26

- Please write doc comments after method names:
```
def sss:
'''
'''
```

**Comment by KonstantinosTsivis** [comment]
Thanks for the feedback. I think I've fixed those issues now, but let me know if I missed something. More specifically:

- fixed pep8 errors
- user can now search with multiple words
- deleted self.api_key or self.url which were not used
- deleted unused imports
- replaced removesource to remove, which is the actual command
- the user can remove now by index, while seeing the already existing sources at the beginning
- simplified news configure
- deleted line about updating sources online

**Comment by appi147** [comment]
Can you please fix tests? I'll see other changes in morning

File: jarviscli/plugins/news.py [hunk 11]
```diff
@@ -1,227 +1,171 @@
...
-try:  # python3
-    import urllib.request
-    import urllib.parse
-    import urllib.error
-except ImportError:  # python2
-    import urllib
 import json
 import webbrowser
-from six.moves import input
 from plugin import plugin, require
```
Comment (appi147) [inline jarviscli/plugins/news.py]:
```suggestion
from six.moves import input
from plugin import plugin, require
```

**Comment by appi147** [comment]
@pnhofmann I need your help.
All of these changes looks good to me.
But this has led to test_news.py to fail. When you change n.get_news_json() to n.get_headlines() in test_news.py, how do you pass jarvis??

**Comment by pnhofmann** [comment]
Sorry, should have documented somewhere.

The MockJarvisAPI instance can be accessed as self.jarvis_api.

Additionally just noticed, that in tests/__init__.py line 69 the get_data method should be changed to something like this:

```
       if key not in self.data:
            value = None
        else:
            value = self.data[key]

        self.call_history.record('get_data', (key), value)
        return value
```

**Comment by appi147** [comment]
> Sorry, should have documented somewhere.
> 
> The MockJarvisAPI instance can be accessed simply as self.jarvis_api.

Thanks

> Additionally just noticed, that in tests/**init**.py line 69 the get_data method should be changed to something like this:
> 
> ```
>        if key not in self.data:
>             value = None
>         else:
>             value = self.data[key]
> 
>         self.call_history.record('get_data', (key), value)
>         return value
> ```

I am patching in this PR.

**Comment by appi147** [comment]
@KonstantinosTsivis Have a look if you are satisfied with all changes, I'll merge

**Comment by KonstantinosTsivis** [comment]
Yep they look good to me too I had the same issue of passing jarvis. :)

**Comment by appi147** [comment]
Nice work!!

**Pull request merged**